kpm (un)install global for all platforms

victorhurdugaci opened the pull request:
Please review: @davidfowl @loudej 
Fixes: https://github.com/aspnet/XRE/issues/540 https://github.com/aspnet/XRE/issues/1140

ChengTian commented on src/Microsoft.Framework.PackageManager/Restore/RestoreCommand.cs:
```diff
@@ -136,6 +145,138 @@ public RestoreCommand(IApplicationEnvironment env)
+
+            Reports.Information.WriteLine("{0}, {1}ms elapsed", "Install complete".Green().Bold(), sw.ElapsedMilliseconds);
+
+            PrintDependencyGraph(graphNode, restoreContext.FrameworkName);
```
Not needed anymore. We just removed this: https://github.com/aspnet/XRE/commit/65473b0a6199d9c886f358d9057cc46c35255d82

victorhurdugaci replied:
OK, rebased and removed

ChengTian commented on src/Microsoft.Framework.PackageManager/Restore/RestoreCommand.cs:
```diff
@@ -136,6 +145,138 @@ public RestoreCommand(IApplicationEnvironment env)
+
+            Reports.Information.WriteLine("Installing package {0} {1}", RestorePackageId.Bold(), RestorePackageVersion.Bold());
+
+            var graphNode = await restoreOperations.CreateGraphNode(
```
Feels like a fair fragment of code can be shared between `RestoreForInstall` and `RestoreForProject`. Can we have `GetPackagesToInstall(IEnumerable<RestoreContext> contexts, IEnumerable<LibraryRange> seeds)`, which can be shared by the two methods?

victorhurdugaci replied:
Talked offline and refactoring this is too time consuming to make sense. The code is very similar but extremely different. I tried to refactor this initially and ended up with a method with 8 arguments...

davidfowl commented on src/Microsoft.Framework.Runtime.Common/Impl/PlatformHelper.cs:
```diff
@@ -16,5 +16,14 @@ public static bool IsMono
+            get
+            {
+                // For now assume Windows = not Mono
+                return !IsMono;
```
I don't think any code should be depending on this. Can we remove it?

victorhurdugaci replied:
You mean it shouldn't depend on `IsMono` or on `IsWindows`?

victorhurdugaci commented:
Ping?

davidfowl commented on src/Microsoft.Framework.PackageManager/Install/InstallBuilder.cs:
```diff
@@ -0,0 +1,175 @@
+    {
+        public const string CommandsFolderName = "app";
+
+        private static readonly string[] NotAllowedExportedCommandNames = new string[]
```
BlockedCommandNames

victorhurdugaci replied:
Renamed

davidfowl commented on src/Microsoft.Framework.PackageManager/Install/AppCommandsFolderRepository.cs:
```diff
@@ -0,0 +1,141 @@
+        {
+            _commands = new Dictionary<string, NuGet.PackageInfo>();
+
+            string pathFilter = PlatformHelper.IsWindows ? "*.cmd" : "*.*";
```
We need to do real OS checks for things like this.

victorhurdugaci replied:
Yup, starting a thread with the CLR team because they don't expose the OS version on CoreCLR

davidfowl commented on src/Microsoft.Framework.PackageManager/Install/InstallBuilder.cs:
```diff
@@ -0,0 +1,175 @@
+            _buildReport.Verbose.WriteLine("Exported command script: " + scriptFileName);
+        }
+
+        private void WriteLinuxScript(string applicationFolder, string commandName)
```
Nix?

victorhurdugaci replied:
Sure

victorhurdugaci commented:
Updated with PR feedback but we have one problem.

@davidfowl we disallow some names like `kpm` and `nuget` but the package manager project exports `kpm` :) What should we do?

victorhurdugaci commented:
Problem fixed. Removed `kpm` from the exported commands list and also added a check at install time to prevent someone from installing a command that's not allowed

lodejard commented on src/Microsoft.Framework.PackageManager/Utils/FileOperationUtils.cs:
```diff
@@ -35,21 +36,25 @@ private static void DeleteRecursive(string deletePath)
                 UseShellExecute = false,
                 FileName = "chmod",
-                Arguments = "+x " + scriptPath
+                Arguments = "+x " + file
```
Probably want to file a new issue to deal with paths-with-spaces on this line

victorhurdugaci replied:
I'll just address that in this PR, no need for a new issue

lodejard commented on src/Microsoft.Framework.Runtime/NuGet/Packages/PackageInfo.cs:
```diff
@@ -34,5 +34,23 @@ public IPackage Package
             }
         }
+
+        public override bool Equals(object obj)
```
Why is the PackageInfo used as an equatable object?

lodejard commented on src/Microsoft.Framework.PackageManager/Install/CommandNameValidator.cs:
```diff
@@ -0,0 +1,30 @@
+            "klr",
+            "kpm",
+            "kvm",
+            "nuget"
```
weren't we going to not create commands for web, test, and run?

lodejard commented on src/Microsoft.Framework.PackageManager/Install/InstallBuilder.cs:
```diff
@@ -0,0 +1,165 @@
+                    .Red());
+
+                return false;
+            }
```
Just skip run, web, test commands - don't need to throw

lodejard commented on src/Microsoft.Framework.PackageManager/Restore/RestoreCommand.cs:
```diff
@@ -136,6 +145,136 @@ public RestoreCommand(IApplicationEnvironment env)
             }
         }
 
+        private async Task<bool> RestoreForInstall(string packagesDirectory)
```
Why was this broken out into RestoreForInstall? Wasn't it part of a single RestoreForProject earlier?

victorhurdugaci replied:
No, it was not. Also, it makes sense to have it as a different `Restore` because there is no project in the context of install

lodejard commented on src/Microsoft.Framework.PackageManager/Utils/PathUtilities.cs:
```diff
@@ -0,0 +1,69 @@
+
+        private static string FindRuntimeHome()
+        {
+            var runtimeHome = Environment.GetEnvironmentVariable("KRE_HOME");
```
Don't use KRE_HOME or KRE_GLOBAL_PATH for install. The install should always be to the user profile `~/.k/bin` irrespective of any of the other settings

lodejard commented on src/Microsoft.Framework.PackageManager/Program.cs:
```diff
@@ -444,6 +415,65 @@ public int Main(string[] args)
                 });
             });
 
+            app.Command("install-g", c =>
```
Change to ("commands") with subcommand ("add")

victorhurdugaci replied:
When we talked a few ago, you mentioned that you would prefer to have distinct commands `install` and `uninstall` for these commands (https://github.com/aspnet/XRE/issues/1135). I can go either way, just let me know if which one you prefer. Btw, I think `commands install` is better than `commands add`

lodejard commented on src/Microsoft.Framework.PackageManager/Program.cs:
```diff
@@ -444,6 +415,65 @@ public int Main(string[] args)
+                });
+            });
+
+            app.Command("uninstall-g", c =>
```
Change to "commands" with subcommand "remove"

victorhurdugaci commented:
@lodejard updated with changes to address your feedback

victorhurdugaci commented:
@lodejard updated command names:

`kpm commands install ...`
`kpm commands uninstall ...`
